src/Features/Core/Portable/Common/UpdatedEventArgs.cs

```diff
@@ -26,12 +28,19 @@ internal class UpdatedEventArgs : EventArgs
         /// </summary>
         public DocumentId DocumentId { get; }
 
-        public UpdatedEventArgs(object id, Workspace workspace, ProjectId projectId, DocumentId documentId)
+        /// <summary>
+        /// The SourceText used to create these Diagnostics.  Only around if the Document was open
+        /// at the time the diagnostics were created.
+        /// </summary>
+        public SourceText OpenSourceText { get; }
```

Comment:
who else except tagger will care about this? if this is something only tagger care then tagger should have this data.

Reply:
Anyone who wants to know "what are these spans valid against".  Right now we just say things like "there is an error in document 'Foo.cs' at span 1-5".  But that is lossy data.  Clients who get this data can't properly know if that span is accurate.  For example, if someone wanted to navigate there, it could put the caret in the wrong place.  

--

I would turn it around.  Right now we lose data, assuming that no one wants it.  But i think that's wrong.  I think people care greatly about the spans for diagnostics, especially knowing where they actually are.

Reply:
if we want to tell uses what version data is created against, then we should use SolutionVersion instead of OpenSourceText.

if tagger wants accurate data, it should use what LB uses (http://index/?leftProject=Microsoft.CodeAnalysis.Features&leftSymbol=dvig2bzttnff&file=Diagnostics%5CIDiagnosticAnalyzerService.cs&line=36)

error list and tagger originally choose what it is using right now accepting data can be wrong but will be fixed up at the end. scarifying accuracy over performance. we we value accuracy more than we should move to API that supports that rather than hack around.

Reply:
> if tagger wants accurate data, it should use what LB use

What is LB?

In the other thread, you said that doing this would be very expensive.  Is that no longer hte case?

Reply:
> if we value accuracy more than we should move to API that supports that rather than hack around.

I do not see what is a hack about keeping track of the SourceText at creation time.  It is legitimately the accurate data.

Reply:
tagging @mavasani for his thoughts here.  We already have the data for open files.  Why is it a hack to use it?

Reply:
well. for me, it is weird, a service that telling you about solution suddenly knows about a file is opened or not.
..
> LB
is light bulb.
..
 mentioned  (http://index/?leftProject=Microsoft.CodeAnalysis.Features&leftSymbol=dvig2bzttnff&file=Diagnostics%5CIDiagnosticAnalyzerService.cs&line=36) since you keep saying DiagnosticService is weird since it doesn't behave same as other. sure it can behave same as other. it is just expensive. so it behave differently. and you are asking why behave different, so I am giving you away to behave same.